README
NAME
    App::perlhl - application class for syntax highlighting Perl source code

VERSION
    version 0.001

SYNOPSIS
        use App::perlhl;
        App::perlhl->new({})->run({}, \@ARGV);

DESCRIPTION
    App::perlhl is the application class backing perlhl.

METHODS
  new
    This instantiates a new App::perlhl object. It takes a hashref of
    options:

    *   html

        If true, the output will be an HTML fragment suitable for publishing
        as part of a web page. NOTE: In the future, this might output a
        whole valid document.

    The default is to output ANSI colour codes suitable for printing to any
    reasonable shell or terminal (which probably means you have the one
    that'll break -- well it works on mine, so neener neener).

  run
    Unsurprisingly, this runs the application. The method takes a hashref of
    options, and an arrayref of filenames to highlight. If there are no
    filenames, defaults to "STDIN".

  Options
    *   version

        If present, the application will print version data and exit.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see <http://search.cpan.org/dist/App-perlhl/>.

    The development version lives at <http://github.com/doherty/App-perlhl>
    and may be cloned from <git://github.com/doherty/App-perlhl.git>.
    Instead of sending patches, please fork this project using the standard
    git and github infrastructure.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://github.com/doherty/App-perlhl/issues>.

AUTHOR
    Mike Doherty <doherty@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Mike Doherty.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Revision history for Perl module App-perlhl

{{$NEXT}}

0.001     2011-02-01
    * Initial experimental implementation


lib/App/perlhl.pm
use strict;
use warnings;
use 5.010100;

package App::perlhl;
# ABSTRACT: application class for syntax highlighting Perl source code

use autodie qw(open close);
use Carp ();
use Syntax::Highlight::Perl::Improved 1.01 ();
use Term::ANSIColor 3.00 ();

=head1 SYNOPSIS

    use App::perlhl;
    App::perlhl->new({})->run({}, \@ARGV);

=head1 DESCRIPTION

B<App::perlhl> is the application class backing L<perlhl>.

=head1 METHODS

=head2 new

This instantiates a new App::perlhl object. It takes a hashref
of options:

=over 4

=item * html

If true, the output will be an HTML fragment suitable for publishing as part
of a web page. B<NOTE:> In the future, this might output a whole valid document.

=back

The default is to output ANSI colour codes suitable for printing to any
reasonable shell or terminal (which probably means you have the one that'll
break -- well it works on mine, so neener neener).

=cut

sub new {
    my $class = shift;
    my $opts  = shift;

    my $formatter = Syntax::Highlight::Perl::Improved->new();
    if ($opts->{html}) {
        my $color_table = {
            'Variable_Scalar'   => 'color:#080;',
            'Variable_Array'    => 'color:#f70;',
            'Variable_Hash'     => 'color:#80f;',
            'Variable_Typeglob' => 'color:#f03;',
            'Subroutine'        => 'color:#980;',
            'Quote'             => 'color:#00a;',
            'String'            => 'color:#00a;',
            'Comment_Normal'    => 'color:#069;font-style:italic;',
            'Comment_POD'       => 'color:#014;font-family:garamond,serif;font-size:11pt;',
            'Bareword'          => 'color:#3A3;',
            'Package'           => 'color:#900;',
            'Number'            => 'color:#f0f;',
            'Operator'          => 'color:#000;',
            'Symbol'            => 'color:#000;',
            'Keyword'           => 'color:#000;',
            'Builtin_Operator'  => 'color:#300;',
            'Builtin_Function'  => 'color:#001;',
            'Character'         => 'color:#800;',
            'Directive'         => 'color:#399;font-style:italic;',
            'Label'             => 'color:#939;font-style:italic;',
            'Line'              => 'color:#000;',
        };
        # HTML escapes.
        $formatter->define_substitution('<' => '&lt;',
                                        '>' => '&gt;',
                                        '&' => '&amp;');

        # install the formats set up above
        while ( my($type, $style) = each %{$color_table} ) {
            $formatter->set_format($type, [ qq{<span style="$style">}, qq{</span>} ]);
        }
    }
    else {
        my $color_table = { # Readability is not so good -- play with it more
            'Bareword'          => 'bright_green',
            'Builtin_Function'  => 'blue',
            'Builtin_Operator'  => 'bright_red',
            'Character'         => 'bold bright_red',
            'Comment_Normal'    => 'bright_blue',
            'Comment_POD'       => 'bright_black',
            'Directive'         => 'bold bright_black',
            'Keyword'           => 'white',
            'Label'             => 'bright_magenta',
            'Line'              => 'white',
            'Number'            => 'bright_red',
            'Operator'          => 'white',
            'Package'           => 'bold bright_red',
            'Quote'             => 'blue',
            'String'            => 'blue',
            'Subroutine'        => 'yellow',
            'Symbol'            => 'white',
            'Variable_Array'    => 'cyan',
            'Variable_Hash'     => 'magenta',
            'Variable_Scalar'   => 'green',
            'Variable_Typeglob' => 'bright_red',
        };

        # install the formats set up above
        while ( my ( $type, $style ) = each %{$color_table} ) {
            $formatter->set_format($type, [ Term::ANSIColor::color($style), Term::ANSIColor::color('reset') ]);
        }
    }

    return bless { formatter => $formatter }, $class;
}

=head2 run

Unsurprisingly, this runs the application. The method takes a hashref of options,
and an arrayref of filenames to highlight. If there are no filenames, defaults to
C<STDIN>.

=head2 Options

=over 4

=item * version

If present, the application will print version data and exit.

=back

=cut

sub run {
    my $self = shift;
    my $opts = shift;
    my $argv = shift;

    if ($opts->{version}) {
        require File::Basename;
        my $this = File::Basename::basename($0); # __PACKAGE__?
        my $this_ver = (defined __PACKAGE__->VERSION ? __PACKAGE__->VERSION : 'dev');
        say "$this version $this_ver" and exit;
    }

    if (@$argv) {
        foreach my $filename (@$argv) {
            open my $in, '<', $filename;
            # Use a separate object for each file - otherwise,
            # highlighting for anything after the first file
            # will be suboptimal.
            my $formatter = $self->{formatter}->new();
            while (<$in>) {
                print $formatter->format_string;
            }
        }
    }
    else {
        while (<STDIN>) {
            print $self->{formatter}->format_string;
        }
    }
}

1;


lib/App/perl/highlight.pm
use perl5i::2;

use Syntax::Highlight::Perl::Improved;

package App::perl::highlight;
# ABSTRACT: Syntax highlight your source code

sub new {
    my $class = shift;

    my $self = {};
    return bless $self, $class;
}

# perl -MText::Highlight -E '$h=Text::Highlight->new(ansi=>1); my $text=do{local $/; open my $fh, "<", $ARGV[0]; <$fh>}; say $h->highlight("Perl", $text);'
sub run {
    my $self = shift;
    my $opts = shift;
    my $args = shift;

    if ($opts->{version}) {
        require File::Basename;
        my $this = File::Basename::basename($0); # __PACKAGE__?
        my $this_ver = (defined __PACKAGE__->VERSION ? __PACKAGE__->VERSION : 'dev');
        say "$this version $this_ver" and exit;
    }

    my $formatter = Syntax::Highlight::Perl::Improved->new();
    if ($opts->{html}) {
        my $color_table = {
            'Variable_Scalar'   => 'color:#080;',
            'Variable_Array'    => 'color:#f70;',
            'Variable_Hash'     => 'color:#80f;',
            'Variable_Typeglob' => 'color:#f03;',
            'Subroutine'        => 'color:#980;',
            'Quote'             => 'color:#00a;',
            'String'            => 'color:#00a;',
            'Comment_Normal'    => 'color:#069;font-style:italic;',
            'Comment_POD'       => 'color:#014;font-family:garamond,serif;font-size:11pt;',
            'Bareword'          => 'color:#3A3;',
            'Package'           => 'color:#900;',
            'Number'            => 'color:#f0f;',
            'Operator'          => 'color:#000;',
            'Symbol'            => 'color:#000;',
            'Keyword'           => 'color:#000;',
            'Builtin_Operator'  => 'color:#300;',
            'Builtin_Function'  => 'color:#001;',
            'Character'         => 'color:#800;',
            'Directive'         => 'color:#399;font-style:italic;',
            'Label'             => 'color:#939;font-style:italic;',
            'Line'              => 'color:#000;',
        };
        # HTML escapes.
        $formatter->define_substitution('<' => '&lt;',
                                        '>' => '&gt;',
                                        '&' => '&amp;');

        # install the formats set up above
        while ( my ( $type, $style ) = each %{$color_table} ) {
            $formatter->set_format( $type, [ qq{<span style="$style">}, qq{</span>} ] );
        }
    }
    else {
        require Term::ANSIColor;
        Carp::croak 'This requires Term::ANSIColor 3.0'
            unless $Term::ANSIColor::VERSION >= 3.00;

        my $color_table = { # Readability is not so good -- play with it more
            'Bareword'          => 'bright_green',
            'Builtin_Function'  => 'blue',
            'Builtin_Operator'  => 'bright_red',
            'Character'         => 'bold bright_red',
            'Comment_Normal'    => 'bright_blue',
            'Comment_POD'       => 'bright_black',
            'Directive'         => 'bold bright_black',
            'Keyword'           => 'white',
            'Label'             => 'bright_magenta',
            'Line'              => 'white',
            'Number'            => 'bright_red',
            'Operator'          => 'white',
            'Package'           => 'bold bright_red',
            'Quote'             => 'blue',
            'String'            => 'blue',
            'Subroutine'        => 'yellow',
            'Symbol'            => 'white',
            'Variable_Array'    => 'cyan',
            'Variable_Hash'     => 'magenta',
            'Variable_Scalar'   => 'green',
            'Variable_Typeglob' => 'bright_red',
        };

        # install the formats set up above
        while ( my ( $type, $style ) = each %{$color_table} ) {
            $formatter->set_format( $type, [ Term::ANSIColor::color($style), Term::ANSIColor::color('reset') ] );
        }
    }

    while (<STDIN>) {
        print $formatter->format_string;
    }
}


Makefile.PL

use strict;
use warnings;

BEGIN { require 5.010100; }

use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'application class for syntax highlighting Perl source code',
  'AUTHOR' => 'Mike Doherty <doherty@cpan.org>',
  'BUILD_REQUIRES' => {
    'File::Find' => '0',
    'File::Temp' => '0',
    'Test::More' => '0',
    'Test::Output' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'App-perlhl',
  'EXE_FILES' => [
    'bin/perlhl'
  ],
  'LICENSE' => 'perl',
  'NAME' => 'App::perlhl',
  'PREREQ_PM' => {
    'Carp' => '0',
    'File::Basename' => '0',
    'Getopt::Long' => '0',
    'Pod::Usage' => '0',
    'Syntax::Highlight::Perl::Improved' => '1.01',
    'Term::ANSIColor' => '3.00',
    'autodie' => '0'
  },
  'VERSION' => '0.002',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



